File: src/action_tutorials_cpp/src/fibonacci_action_client.cpp
#include <chrono>
#include <memory>
#include <vector>
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>
#include "action_tutorials_interfaces/action/fibonacci.hpp"

using namespace std::chrono_literals;
using Fibonacci = action_tutorials_interfaces::action::Fibonacci;

class FibonacciActionClient : public rclcpp::Node {
public:
    FibonacciActionClient(int timeout_seconds, double sleep_seconds)
        : Node("fibonacci_action_client"),
          timeout_seconds_(timeout_seconds),
          sleep_duration_(std::chrono::duration<double>(sleep_seconds)),
          start_time_(this->now()) {
        
        action_client_ = rclcpp_action::create_client<Fibonacci>(this, "fibonacci");
    }

    void send_goals(int order) {
        while (rclcpp::ok() && (this->now() - start_time_).seconds() < timeout_seconds_) {
            RCLCPP_INFO(this->get_logger(), "Sending goal with order %d", order);

            auto goal_msg = Fibonacci::Goal();
            goal_msg.order = order;

            if (!action_client_->wait_for_action_server(std::chrono::seconds(5))) {
                RCLCPP_ERROR(this->get_logger(), "Action server not available!");
                return;
            }

            rclcpp_action::Client<Fibonacci>::SendGoalOptions options;
            options.feedback_callback =
                std::bind(&FibonacciActionClient::feedback_callback, this, std::placeholders::_1, std::placeholders::_2);
            options.result_callback =
                std::bind(&FibonacciActionClient::result_callback, this, std::placeholders::_1);

            auto goal_handle_future = action_client_->async_send_goal(goal_msg, options);

            // Wait for the goal to be accepted synchronously
            rclcpp::spin_until_future_complete(shared_from_this(), goal_handle_future, std::chrono::seconds(5));
            auto goal_handle = goal_handle_future.get();
            if (!goal_handle) {
                RCLCPP_ERROR(this->get_logger(), "Goal was rejected by the server.");
                return;
            }

            // Now, we will wait until the result is processed via the result callback
            RCLCPP_INFO(this->get_logger(), "Goal accepted, waiting for result...");
            
            // The result callback will signal when the result is available
            // We block the sending of the next goal until this is processed
            goal_result_ready_ = false;
            while (!goal_result_ready_) {
                rclcpp::spin_some(shared_from_this());  // Spin to process result callback
                std::this_thread::sleep_for(std::chrono::milliseconds(100)); // Small sleep to avoid high CPU usage
            }

        }

        RCLCPP_INFO(this->get_logger(), "Timeout reached, stopping client.");
    }

private:
    rclcpp_action::Client<Fibonacci>::SharedPtr action_client_;
    rclcpp::Time start_time_;
    int timeout_seconds_;
    std::chrono::duration<double> sleep_duration_;
    bool goal_result_ready_ = false;

    void feedback_callback(
        std::shared_ptr<rclcpp_action::ClientGoalHandle<Fibonacci>>,
        const std::shared_ptr<const Fibonacci::Feedback> feedback) {
        RCLCPP_INFO(this->get_logger(), "Feedback received:");
        for (auto num : feedback->partial_sequence) {
            RCLCPP_INFO(this->get_logger(), "%d", num);
        }
    }

    void result_callback(
        const rclcpp_action::ClientGoalHandle<Fibonacci>::WrappedResult & result) {
        switch (result.code) {
            case rclcpp_action::ResultCode::SUCCEEDED:
                RCLCPP_INFO(this->get_logger(), "Result received:");
                for (auto num : result.result->sequence) {
                    RCLCPP_INFO(this->get_logger(), "%d", num);
                }
                break;
            case rclcpp_action::ResultCode::ABORTED:
                RCLCPP_ERROR(this->get_logger(), "Goal was aborted");
                break;
            case rclcpp_action::ResultCode::CANCELED:
                RCLCPP_ERROR(this->get_logger(), "Goal was canceled");
                break;
            default:
                RCLCPP_ERROR(this->get_logger(), "Unknown result code");
                break;
        }

        // Set the flag to signal the result is ready
        goal_result_ready_ = true;
    }
};

int main(int argc, char ** argv) {
    rclcpp::init(argc, argv);

    if (argc != 3) {
        std::cerr << "Usage: ros2 run your_package your_client_node <timeout_seconds> <sleep_seconds>" << std::endl;
        return 1;
    }

    int timeout_seconds = std::stoi(argv[1]);
    double sleep_seconds = std::stod(argv[2]);

    auto client = std::make_shared<FibonacciActionClient>(timeout_seconds, sleep_seconds);
    client->send_goals(10);

    rclcpp::shutdown();
    return 0;
}

File: src/action_tutorials_py/action_tutorials_py/fibonacci_action_client.py
import sys
import time
import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node

from action_tutorials_interfaces.action import Fibonacci


class FibonacciActionClient(Node):

    def __init__(self, timeout_seconds, sleep_seconds):
        super().__init__('fibonacci_action_client')

        self.sleep_duration = sleep_seconds
        self.timeout = timeout_seconds
        self.start_time = time.time()

        self._action_client = ActionClient(self, Fibonacci, 'fibonacci')

        self._timeout_occurred = False
        self._goal_handle = None
        self._timeout_timer = None

    def send_goal_and_wait(self, order):
        self.get_logger().info(f"Sending goal with order {order}")
        goal_msg = Fibonacci.Goal()
        goal_msg.order = order

        #if not self._action_client.wait_for_server(timeout_sec=5.0):
        if not self._action_client.wait_for_server():
            self._node.get_logger().error('Action server not available!')
            return

        self._send_goal_future = self._action_client.send_goal_async(
            goal_msg, feedback_callback=self.feedback_callback
        )

        self._timeout_timer = self.create_timer(self.timeout, self.on_timeout)

        #rclpy.spin_until_future_complete(self, self._send_goal_future, timeout_sec=5.0)
        rclpy.spin_until_future_complete(self, self._send_goal_future)

        self._goal_handle = self._send_goal_future.result()

        if not self._goal_handle.accepted:
            self.get_logger().info('Goal was not accepted.')
            if self._timeout_timer:
                self._timeout_timer.cancel()
            return False

        self.get_logger().info("Goal accepted, waiting for result.")
        self._get_result_future = self._goal_handle.get_result_async()
        #rclpy.spin_until_future_complete(self, self._get_result_future, timeout_sec=5.0)
        rclpy.spin_until_future_complete(self, self._get_result_future)

        # If the result is obtained or timeout occurs, stop waiting
        result = self._get_result_future.result()
        if result:
            self.get_logger().info(f"Result: {result.result.sequence}")
        else:
            self.get_logger().info("Result not received due to timeout.")

            #sys.exit(0)

        # Cancel the timeout timer if we received the result or feedback
        if self._timeout_timer:
            self._timeout_timer.cancel()

        return True

    def feedback_callback(self, feedback_msg):
        feedback = feedback_msg.feedback
        self.get_logger().info(f'Received feedback: {feedback.partial_sequence}')

        # If feedback is received, cancel the timeout timer
        if self._timeout_timer:
            self._timeout_timer.cancel()

    def on_timeout(self):
        self.get_logger().warning('Action timed out.')
        self._timeout_occurred = True

        if self._goal_handle:
            self._goal_handle.cancel_goal_async()

        # Exit the action client logic
        self.get_logger().info('Goal cancelled due to timeout.')

        # Optionally, cancel any other action processes, like result waiting
        if self._get_result_future and not self._get_result_future.done():
            self._get_result_future.cancel()

        # sys.exit(0)

def main(args=None):
    rclpy.init()

    if len(sys.argv) != 3:
        print("Usage: ros2 run your_package your_server_node <timeout_seconds> <sleep_seconds>")
        return

    try:
        timeout_seconds = int(sys.argv[1])
        sleep_seconds = float(sys.argv[2])
    except ValueError:
        print("Please provide valid integer for timeout and float for sleep.")
        return

    action_client = FibonacciActionClient(timeout_seconds, sleep_seconds)

    while rclpy.ok():
        if not action_client.send_goal_and_wait(10):
            break  # Exit loop if timeout is reached
        time.sleep(action_client.sleep_duration)


if __name__ == '__main__':
    main()